conda_libmamba_solver/shards.py

```diff
@@ -0,0 +1,341 @@
+        """
+        return {package: self.fetch_shard(package, session) for package in packages}
+
+    def build_repodata(self):
+        """
+        Return monolithic repodata including all visited shards.
+        """
+        repodata = self.repodata.copy()
+        repodata.update({"packages": {}, "packages.conda": {}})
+        for package, shard in self.visited.items():
+            if shard is None:
+                continue  # recorded visited but not available shards
+            for package_group in ("packages", "packages.conda"):
+                repodata[package_group].update(shard[package_group])
+        return repodata
+
+
+FETCHED_THIS_PROCESS = set()
+
+
+class Shards(ShardLike):
+    def __init__(self, shards_index: ShardsIndex, url: str):
+        """
+        Args:
+            shards_index: raw parsed msgpack dict
+            url: URL of repodata_shards.msgpack.zst
+        """
+        self.shards_index = shards_index
+        self.url = url
+
+        self.repodata = {k: v for k, v in self.shards_index.items() if k not in ("shards",)}
+
+        # used to write out repodata subset
+        self.visited: dict[str, Shard | None] = {}
+
+    def __contains__(self, package):
+        return package in self.packages_index
+
+    @property
+    def packages_index(self):
+        return self.shards_index["shards"]
+
+    def shard_url(self, package: str):
+        """
+        Return shard URL for a given package.
+
+        Raise KeyError if package is not in the index.
+        """
+        shard_name = f"{self.packages_index[package].hex()}.msgpack.zst"
+        # "Individual shards are stored under the URL <shards_base_url><sha256>.msgpack.zst"
+        return urljoin(self.url, f"{self.shards_index['info']['shards_base_url']}{shard_name}")
+
+    def fetch_shard(self, package: str, session) -> Shard:
+        """
+        Fetch an individual shard.
+
+        Raise KeyError if package is not in the index.
+        """
+
+        shard_url = self.shard_url(package)
+        raw_shard = session.get(shard_url).content
+        # ensure it is real msgpack+zstd before inserting into cache
+        shard: Shard = msgpack.loads(zstandard.decompress(raw_shard))  # type: ignore
+        self.visited[package] = shard
+        return shard
+
+    def fetch_shards(self, packages: Iterable[str], session) -> dict[str, Shard]:
+        """
+        Return mapping of *package names* to Shard for given packages.
+        """
+        result = {}
+
+        def fetch(s, url, package):
+            if url in FETCHED_THIS_PROCESS:
+                print("Already got", url)
+                raise RuntimeError("Can't fetch same url twice")
+            FETCHED_THIS_PROCESS.add(url)
+            b1 = time.time_ns()
+            data = s.get(url).content
+            e1 = time.time_ns()
+            print(f"Fetch took {(e1 - b1) / 1e9}s", package, url)
+            return (url, package, data)
+
+        packages = sorted(list(packages))
```

Comment:
I don't think you need the list() call, do you?

```suggestion
        packages = sorted(packages)
```